### ISSUE TITLE
Fix compatibility with Firefox

### ISSUE THREAD

**Issue by jurca** [jurca/indexed-db.es6#3]
Firefox incorrectly throws `Error "A request was placed against a transaction which is currently not active, or which is finished."` when trying to place a request on a still active transaction accessed from a different execution stack (for example a `Promise` callback).

Cursor example:

```
  it("should work", (done) => {
    indexedDB.open(DB_NAME).onsuccess = (event) => {
      let db = event.target.result

      let start = Date.now()

      let transaction = db.transaction(STORE1)

      function keepAlive() {
        transaction.objectStore(STORE1).get(0).onsuccess = () => {
          if (Date.now() - start < 50) {
            console.log("keep-alive")
            keepAlive()
          } else {
            done()
          }
        }
      }

      keepAlive()

      console.log("opening")
      let request = transaction.objectStore(STORE1).openCursor()
      console.log(request)
      asPromise(request).then((cursor) => {
        console.log(cursor.value)

        cursor.continue()
        return asPromise(request)
      }).then((cursor) => {
        console.log(cursor.value)

        cursor.continue()
        return asPromise(request)
      }).then((cursor) => {
        console.log(cursor.value)
      }).catch((error) => {
        console.log(error.message)
        console.log(error.stack)
        fail(error.message)
        done()
      })

      function asPromise(request) {
        return new Promise((resolve, reject) => {
          request.onsuccess = () => resolve(request.result)
          request.onerror = () => reject(request.error)
        })
      }
    }
  })
```

Record adding example:

```
  it("should work", (done) => {
    indexedDB.open(DB_NAME).onsuccess = (event) => {
      let db = event.target.result

      let start = Date.now()

      let transaction = db.transaction(STORE1, "readwrite")

      function keepAlive() {
        transaction.objectStore(STORE1).get(0).onsuccess = () => {
          if (Date.now() - start < 50) {
            console.log("keep-alive")
            keepAlive()
          } else {
            done()
          }
        }
      }

      keepAlive()

      console.log("running")
      asPromise(transaction.objectStore(STORE1).add("test1")).then((key) => {
        console.log(key)
        return asPromise(transaction.objectStore(STORE1).add("test2"))
      }).then((key) => {
        console.log(key)
        return asPromise(transaction.objectStore(STORE2).add("test3"))
      }).then((key) => {
        console.log(key)

        done()
      }).catch((error) => {
        console.log(error.message)
        console.log(error.stack)
        fail(error.message)
        done()
      })

      function asPromise(request) {
        return new Promise((resolve, reject) => {
          request.onsuccess = () => resolve(request.result)
          request.onerror = () => reject(request.error)
        })
      }
    }
  })
```

Sending "keep alive" requests keeps the transaction alive (at the cost of additional overhead) but does not fix this issue as shown in the code examples.

Proposed solution: Create a `PromiseSync` alternative to `Promise` that executes callback synchronously within the same call stack, thus allowing the `then` callbacks to execute operations. The `PromiseSync` will be used for in-transaction operations, the rest of the API will return `Promise` instances.

**Comment by jurca** [comment]
The `KeepAlive` and `RequestMonitor` can be removed after this.

**Comment by jurca** [comment]
`PromiseSync` seems to solve many issues that were present when used with Firefox.

Tests currently failing in Firefox:
- `migration/DatabaseVersionMigratorSpec`
  - `it("should execute the callback within the versionchange transaction"`
  - `it("should cancel versionchange transaction on early error"`
- `transaction/ReadOnlyTransactionSpec`
  - `it("should provide promise rejected on error"`
  - `it("should execute error listeners on error"`
- `object-store/AbstractReadOnlyStorageSpec`
  - `it("should allow paging the records using the list method"`
- `object-store/ObjectStoreSpec`
  - `it("should delete records matching a predicate function"`

**Comment by jurca** [comment]
Tests currently failing in Firefox:
- `migration/DatabaseVersionMigratorSpec`
  - `it("should execute the callback within the versionchange transaction"`
  - `it("should cancel versionchange transaction on early error"`

The question is: how to abort a versionchange transaction without locking up the whole IndexedDB in Firefox?

**Comment by jurca** [comment]
Fixed the remaining tests in 28a2290f82a102c54f8bdc4b898589f68899344c
